File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.js
function solution(maps) {
    const map = [];
    for(const m of maps){
        map.push(m.split(''))
    }
    
    let lever
    for(let i=0; i<map.length; i+=1){
        for(let j=0; j<map[i].length; j+=1){
            if(map[i][j]==='L') lever = [i, j]
        }
    }
    const a = bfs(lever, 'S', map);
    const b = bfs(lever, 'E', map);
    if(a<0 || b<0) return -1
    return a+b
}

function bfs(start, target, map){
    const queue = [[...start, 0]];
    const visited = Array.from({length: map.length}, ()=>Array(map[0].length).fill(false))
    visited[start[0]][start[1]] = true
    const xd = [ -1, 1, 0, 0 ]
    const yd = [ 0, 0, -1, 1 ]
    
    while(queue.length){
        const [x, y, dist] = queue.shift();
        
        if(map[x][y]===target) return dist;
        
        for(let i=0; i<4; i+=1){
            const dx = x + xd[i];
            const dy = y + yd[i];
            
            if(
                (dx>=0 && dx<map.length) &&
                (dy>=0 && dy<map[0].length) &&
                map[dx][dy] !== 'X' &&
                !visited[dx][dy]
              ){
                visited[dx][dy] = true;
                queue.push([dx, dy, dist+1])
            }
        }
    }
    return -1
}

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    start=[]
    lever=[]
    exit=[]
    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if(maps[i][j]=='S'):
                start = [i,j]
            if(maps[i][j]=='L'):
                lever = [i,j]
            if(maps[i][j]=='E'):
                exit = [i,j]
    
    a=bfs(start,lever,maps)
    if(a==-1): return -1
    b=bfs(lever,exit,maps)
    if(b==-1): return -1
    return a+b

def bfs(start, target, map):
    queue = deque()
    queue.append(start)
    n = len(map)
    m = len(map[0])
    dir = [
        [0,1],
        [0,-1],
        [1,0],
        [-1,0]
    ]
    visited = []
    for i in range(n):
        visited.append([float('inf')]*m)
    visited[start[0]][start[1]] = 0
    
    while queue:
        cur = queue.popleft()
        
        if map[cur[0]][cur[1]]==target:
            break
        
        for d in dir:
            rn, rc = [cur[0]+d[0], cur[1]+d[1]]
            
            if(rn<0 or rc<0 or rn>=n or rc>=m): continue
            if(map[rn][rc]=='X'): continue
            if(visited[rn][rc]!=float('inf')): continue
            queue.append([rn, rc])
            visited[rn][rc] = visited[cur[0]][cur[1]] + 1
    return visited[target[0]][target[1]] if visited[target[0]][target[1]]!=float('inf') else -1